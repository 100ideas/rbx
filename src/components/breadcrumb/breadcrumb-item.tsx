import classNames from "classnames";
import PropTypes from "prop-types";
import React from "react";

import { forwardRefAs, Generic, HelpersProps } from "../../base";

export interface BreadcrumbItemModifierProps {
  active?: boolean;
}

export type BreadcrumbItemProps = HelpersProps & BreadcrumbItemModifierProps;

export const BreadcrumbItem = forwardRefAs<
  HTMLAnchorElement,
  BreadcrumbItemProps
>(
  ({ active, ...rest }, ref) => (
    <li className={classNames({ "is-active": active })}>
      <Generic ref={ref} {...rest} />
    </li>
  ),
  { as: "a" },
);

BreadcrumbItem.displayName = "Breadcrumb.Item";
BreadcrumbItem.propTypes = {
  active: PropTypes.bool,
};
