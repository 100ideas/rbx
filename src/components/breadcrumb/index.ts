import "./breadcrumb.module.sass";

export { Breadcrumb } from "./breadcrumb";
