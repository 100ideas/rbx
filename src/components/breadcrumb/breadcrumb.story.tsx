import { select } from "@storybook/addon-knobs";
import { storiesOf } from "@storybook/react";
import React from "react";

import Box from "components/box";
import Breadcrumb from "components/breadcrumb";
import { BreadcrumbProps } from "components/breadcrumb/Breadcrumb";

const makeSeparator = () =>
  select(
    "Separator",
    {
      arrow: "arrow",
      bullet: "bullet",
      dot: "dot",
      succeeds: "succeeds",
    },
    "arrow",
  );

const items = [
  {
    name: "Storybook",
    url: "#1",
  },
  {
    name: "Breadcrumb",
    url: "#2",
  },
  {
    active: true,
    name: "Breadcrumb Types",
    url: "#3",
  },
];

storiesOf("Breadcrumb", module)
  .add("Default", () => (
    <div>
      <Box>
        <Breadcrumb separator={makeSeparator()} items={items} />
      </Box>
    </div>
  ))
  .add("Use Custom render Element", () => {
    const Anchor = ({ children, unselectable, ...props }: BreadcrumbProps) => (
      <a className="Others" {...props}>
        {children}
      </a>
    );

    return (
      <div>
        <Box>
          <Breadcrumb
            renderAs={Anchor}
            hrefAttr="href"
            separator={makeSeparator()}
            items={items}
          />
        </Box>
      </div>
    );
  });
