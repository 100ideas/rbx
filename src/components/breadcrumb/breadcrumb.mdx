---
name: Breadcrumb
menu: Components
---

import { Playground, PropsTable } from "docz";

import { Box } from "components/box";
import { Breadcrumb } from "components/breadcrumb";
import { Content } from "components/content";
import { Image } from "components/image";
import { Media } from "components/media";

# Breadcrumb

## Basic usage

export const meta = {
  items: [
    {
      name: "Storybook",
      url: "#1"
    },
    {
      name: "Breadcrumb",
      url: "#2"
    },
    {
      active: true,
      name: "Breadcrumb Types",
      url: "#3"
    }
  ]
};

<Playground>
  <Box>
    <Breadcrumb separator="arrow" items={
      [
        {
          name: "Storybook",
          url: "#1"
        },
        {
          name: "Breadcrumb",
          url: "#2"
        },
        {
          active: true,
          name: "Breadcrumb Types",
          url: "#3"
        }
      ]
    } />

  </Box>
</Playground>

## Properties

<PropsTable of={Box} />
