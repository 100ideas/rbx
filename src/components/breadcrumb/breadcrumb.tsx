import { cx } from "emotion";
import * as React from "react";

import { asExoticComponent } from "@/components/exotic";
import { ModifierProps, modify } from "@/modifiers";

export interface BreadcrumbItemProps {
  url: string;
  active?: boolean;
  name?: React.ReactNode;
}

export type BreadcrumbModifierProps = Partial<{
  align: "right" | "center";
  className: string;
  hrefAttr: string;
  items: BreadcrumbItemProps[];
  separator: "arrow" | "bullet" | "dot" | "succeeds";
  size: "small" | "medium" | "large";
}>;

export type BreadcrumbProps = ModifierProps &
  BreadcrumbModifierProps &
  Partial<Omit<React.ComponentPropsWithoutRef<"a">, "unselectable">>;

// TODO: should split up Breadcrumb -> Breadcrumb & BreadcrumbItem
// this is because the `ref` is passed down to the breadcrumb container
// but the `as` is passed to the bredcrumb item.
// Ergo: the type system expects `as` to be compatible with the `Ref<type>`,
// but it won't be. (the item defaults to an <a>, and the container is fixed as
// a<nav>)
export const Breadcrumb = asExoticComponent<BreadcrumbProps, "a">(
  (props, ref) => {
    const { align, as, hrefAttr, items, separator, size, ...rest } = modify(
      props,
    );
    rest.className = cx("breadcrumb", rest.className, {
      [`has-${separator}-separator`]: separator,
      [`is-${align}`]: align,
      [`is-${size}`]: size,
    });

    return (
      <nav {...rest} ref={ref}>
        <ul>
          {items!.map(item => {
            const itemProps: { [s: string]: BreadcrumbItemProps["url"] } = {};
            if (as === "a") {
              itemProps.href = item.url;
            } else if (typeof hrefAttr === "string") {
              itemProps[hrefAttr] = item.url;
            }
            return (
              <li key={item.url} className={cx({ "is-active": item.active })}>
                {React.createElement(as!, itemProps, item.name)}
              </li>
            );
          })}
        </ul>
      </nav>
    );
  },
  "a",
);

Breadcrumb.defaultProps = Object.assign({ items: [] }, Breadcrumb.defaultProps);
