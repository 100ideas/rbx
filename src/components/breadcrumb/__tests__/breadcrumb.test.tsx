import React from "react";
import renderer from "react-test-renderer";

import { Breadcrumb, BreadcrumbProps } from "../breadcrumb";

describe("Breadcrumb component", () => {
  it("Should be a Breadcrumb", () => {
    const component = renderer.create(
      <Breadcrumb
        items={[
          {
            name: "Home",
            url: "/",
          },
          {
            name: "Section",
            url: "/section",
          },
          {
            active: true,
            name: "Details",
            url: "/detail",
          },
        ]}
      />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  [undefined, "identifier"].map(hrefAttr =>
    it(`Should ${
      hrefAttr ? "" : "not "
    }pass along item.url as hrefAttr prop on 'as' component when hrefAttr is ${
      hrefAttr ? "" : "not "
    }supplied`, () => {
      interface CustomProps {
        identifier?: string;
      }
      const Custom: React.FC<CustomProps> = ({ identifier }) => (
        <div>{identifier}</div>
      );

      const component = renderer.create(
        <Breadcrumb<typeof Custom>
          as={Custom}
          hrefAttr={hrefAttr}
          items={[
            {
              name: "Storybook",
              url: "#1",
            },
            {
              name: "Breadcrumb",
              url: "#2",
            },
            {
              active: true,
              name: "Breadcrumb Types",
              url: "#3",
            },
          ]}
        />,
      );
      expect(component.toJSON()).toMatchSnapshot();
    }),
  );

  [null, "arrow", "dot", "bullet", "succeeds"].map(separator =>
    it(`should use separator ${separator}`, () => {
      const component = renderer.create(
        <Breadcrumb
          separator={separator as BreadcrumbProps["separator"]}
          items={[
            {
              name: "Storybook",
              url: "#1",
            },
            {
              name: "Breadcrumb",
              url: "#2",
            },
            {
              active: true,
              name: "Breadcrumb Types",
              url: "#3",
            },
          ]}
        />,
      );
      expect(component.toJSON()).toMatchSnapshot();
    }),
  );

  it("Should use inline style and custom size", () => {
    const component = renderer.create(
      <Breadcrumb
        style={{ marginTop: 10 }}
        size="large"
        items={[
          {
            name: "Home",
            url: "/",
          },
          {
            name: "Section",
            url: "/section",
          },
          {
            active: true,
            name: "Details",
            url: "/detail",
          },
        ]}
      />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
