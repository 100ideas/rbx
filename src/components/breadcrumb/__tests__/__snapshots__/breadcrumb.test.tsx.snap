// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb component should be a Breadcrumb 1`] = `
<nav
  className="breadcrumb"
>
  <ul>
    <li
      className=""
    >
      <a
        href="/"
      >
        Home
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/section"
      >
        Section
      </a>
    </li>
    <li
      className="is-active"
    >
      <a
        href="/detail"
      >
        Details
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Breadcrumb component should not pass along item.url as hrefAttr prop on 'as' component when hrefAttr is not supplied 1`] = `
<nav
  className="breadcrumb"
>
  <ul>
    <li
      className=""
    >
      <div />
    </li>
    <li
      className=""
    >
      <div />
    </li>
    <li
      className="is-active"
    >
      <div />
    </li>
  </ul>
</nav>
`;

exports[`Breadcrumb component should pass along item.url as hrefAttr prop on 'as' component when hrefAttr is supplied 1`] = `
<nav
  className="breadcrumb"
>
  <ul>
    <li
      className=""
    >
      <div>
        #1
      </div>
    </li>
    <li
      className=""
    >
      <div>
        #2
      </div>
    </li>
    <li
      className="is-active"
    >
      <div>
        #3
      </div>
    </li>
  </ul>
</nav>
`;

exports[`Breadcrumb component should use inline style and custom size 1`] = `
<nav
  className="breadcrumb is-large"
  style={
    Object {
      "marginTop": 10,
    }
  }
>
  <ul>
    <li
      className=""
    >
      <a
        href="/"
      >
        Home
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="/section"
      >
        Section
      </a>
    </li>
    <li
      className="is-active"
    >
      <a
        href="/detail"
      >
        Details
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Breadcrumb component should use separator arrow 1`] = `
<nav
  className="breadcrumb has-arrow-separator"
>
  <ul>
    <li
      className=""
    >
      <a
        href="#1"
      >
        Storybook
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="#2"
      >
        Breadcrumb
      </a>
    </li>
    <li
      className="is-active"
    >
      <a
        href="#3"
      >
        Breadcrumb Types
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Breadcrumb component should use separator bullet 1`] = `
<nav
  className="breadcrumb has-bullet-separator"
>
  <ul>
    <li
      className=""
    >
      <a
        href="#1"
      >
        Storybook
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="#2"
      >
        Breadcrumb
      </a>
    </li>
    <li
      className="is-active"
    >
      <a
        href="#3"
      >
        Breadcrumb Types
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Breadcrumb component should use separator dot 1`] = `
<nav
  className="breadcrumb has-dot-separator"
>
  <ul>
    <li
      className=""
    >
      <a
        href="#1"
      >
        Storybook
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="#2"
      >
        Breadcrumb
      </a>
    </li>
    <li
      className="is-active"
    >
      <a
        href="#3"
      >
        Breadcrumb Types
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Breadcrumb component should use separator null 1`] = `
<nav
  className="breadcrumb"
>
  <ul>
    <li
      className=""
    >
      <a
        href="#1"
      >
        Storybook
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="#2"
      >
        Breadcrumb
      </a>
    </li>
    <li
      className="is-active"
    >
      <a
        href="#3"
      >
        Breadcrumb Types
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Breadcrumb component should use separator succeeds 1`] = `
<nav
  className="breadcrumb has-succeeds-separator"
>
  <ul>
    <li
      className=""
    >
      <a
        href="#1"
      >
        Storybook
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="#2"
      >
        Breadcrumb
      </a>
    </li>
    <li
      className="is-active"
    >
      <a
        href="#3"
      >
        Breadcrumb Types
      </a>
    </li>
  </ul>
</nav>
`;
